packages/matter.js/src/cluster/server/AttributeServer.ts

```diff
@@ -64,21 +75,73 @@ export class FixedAttributeServer<T> extends BaseAttributeServer<T> {
         if (version !== undefined) {
             throw new Error("Version is not supported on fixed attributes");
         }
-        this.validator(value, this.name);
+        this.schema.validate(value);
         this.value = value;
     }
 }
 
+/**
+ * Base class for attribute servers that back mutable (non-fixed) attributes.
+ */
+export abstract class MutableAttributeServer<T> extends BaseAttributeServer<T> {
+    set(value: T, session?: Session<MatterDevice>) {
+        // TODO - ACL enforcement
+
+        this.setFabricIndex(value, session?.getAccessingFabric().fabricIndex, this.schema);
+
+        this.setRemote(value, session);
+    }
+
+    /**
+     * Set an attribute value in the context of a remote request.
+     */
+    abstract setRemote(value: T, session?: Session<MatterDevice>): void;
+
+    /**
+     * Set the value for the global "FabricIndex" attribute.
+     */
+    private setFabricIndex(value: any, fabricIndex: FabricIndex | undefined, schema: TlvSchema<any>) {
+        if (schema instanceof ArraySchema) {
+            if (Array.isArray(value)) {
+                for (const item of value) {
+                    this.setFabricIndex(item, fabricIndex, schema.elementSchema);
+                }
+            }
+        } else if (schema instanceof NullableSchema) {
+            this.setFabricIndex(value, fabricIndex, schema.schema);
+        } else if (schema instanceof ObjectSchema) {
+            if (typeof schema !== "object") {
+                return;
+            }
+
+            for (const k in schema.fieldDefinitions) {
```

Comment:
maybe a find is better here? Or at least a break after id found because dplucate IDs should not happen :-)

Reply:
It continues the loop because if it's not 0xfe it recurses into subfields

Reply:
Ahhh yes. Kk